### PR TITLE
Make it easier to integrate alt string / buffer types

### PR DESCRIPTION
**Pull request by kmcallister**
I'm using this in https://github.com/kmcallister/tendril.

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -318,34 +326,42 @@ pub trait Encoding {
     /// or a failure to return the error.
     #[stable]
     fn decode(&self, input: &[u8], trap: DecoderTrap) -> Result<String, Cow<'static, str>> {
+        let mut ret = String::new();
+        self.decode_to(input, trap, &mut ret).map(|_| ret)
+    }
+
+    /// Decode into a `StringWriter`.
```
Comment (lifthrasiir) [inline src/types.rs]: While I'm open to the convenience methods, I think this should clarify that `input` cannot be partial input. That was why I didn't add such methods so far (the proper incremental interface should be used instead).

File: src/types.rs [hunk 2]
```diff
@@ -279,34 +279,42 @@ pub trait Encoding {
     /// or a failure to return the error.
     #[stable]
     fn encode(&self, input: &str, trap: EncoderTrap) -> Result<Vec<u8>, Cow<'static, str>> {
+        let mut ret = Vec::new();
+        self.encode_to(input, trap, &mut ret).map(|_| ret)
+    }
+
+    /// Encode into a `ByteWriter`.
```
Comment (lifthrasiir) [inline src/types.rs]: Same, but not mandatory. (There is no practical encoding with a special encoding for two or more Unicode scalar values without exception.)

**Comment by kmcallister** [comment]
I added a warning on `decode_to`.

**Comment by lifthrasiir** [comment]
Thank you!

**Pull request merged**